### ISSUE TITLE
canPlayType method on the player

### ISSUE THREAD

**Issue by gkatsev** [videojs/video.js#2701]
It seems we never had a `canPlayType` on the player. Since is super useful to have with source handlers so that we can know whether a player can play something. The player level-method will delegate to the techs which will in-turn delegate to any source handlers and it'll return the first truthy value.
For example, contrib-hls will allows us to return true for `application/x-mpegurl` or `application/vnd.apple.mpegurl`.
```js
player.canPlayType('application/x-mpegurl');
```
Previously, you had to check directly whether a particular tech exists but that isn't easily done with source handlers.

Once we have the above, we could even come up with a scheme for custom techs like @eXon's youtube tech.
```js
player.canPlayType('vjs-tech/youtube')
```
Though, this is definitely a stretch goal.

**Comment by imbcmdth** [comment]
Surely, you mean `application/youtube`!

**Comment by gkatsev** [comment]
`application/x-youtube`. We probably shouldn't do anything too official.

**Comment by eXon** [comment]
I believe tech all tech are having something like this:

```javascript
videojs.Youtube.isSupported = function() {
    return true;
  };
```

This shouldn't be too hard to do.

**Comment by gkatsev** [comment]
Nope, not too hard. Just can't believe it didn't exist before. And the point is that we shouldn't have people go find if a particular tech or source handler is available and then call its `isSupported` method. Plus, it makes the player more inline with the html5 API. :)

**Comment by nickygerritsen** [comment]
Yes we would be very happy with this I think :).

We want to provide both adaptive content (HLS, MPEG-DASH) and progressive content with quality switching (using a quality dropdown like Youtube) to the player and prefer adaptive content, but only if it can be played without using Flash. I think this could be done with this function, but maybe only if it returns whether the playback is "preferred" or something?

I.e. if the player has two techs, X and Y, and source A can be played by tech Y only, while source B can be played by both techs, I'd like to play source Y. Is this something that is possible? Perhaps we should let the function return a number instead of a boolean, where "-1" means "can not play" and any number N >=0 means "it can be played and has 'priority' N"? Or is this a strange idea?

**Comment by gkatsev** [comment]
I think your suggestion is an enhancement to the tech/source selection algorithm as opposed to an addition to the canPlayType method. Ideally, I'd want to keep this method in line with the native html5 method unless there's really a good reason to do otherwise (like maybe just returning a boolean as opposed to a string enum). I guess right now, assuming html5 `maybe` and `probably`, `canPlayType` should probably return the "best" answer. So, return the first `probably`, but if we receive a `maybe`, it would need to try other techs to see if they can return a `probably` instead. Though, our tech selection algorithm doesn't currently take that into account anyway.

The tech priority thing is interesting though because it matters only in terms of something like playlists or multiple sources in one player, though, videojs/html5 is kind of only set up to work with single sources. Ideally, switching techs should not be an issue but that's never true.

**Comment by nickygerritsen** [comment]
But in the current setup the `selectSource` function exactly does what I suggest right?

I.e. if I pass HLS and progressive and HTML5 only supports progressive in the current browser, the progressive gets chosen. Because the loop goes through every source for each tech, so "higher" techs have priority over "lower" ones.

So for me what I want is something that corresponds to that. Although I could of course just pass all adaptive video's and one progressive one to `player.selectSource` and inspect what the return value is. But that feels a bit ugly.

And yes, we want to use it for multiple sources + playlist in one player, which we already got working quite easily with only progressive video's.

**Comment by gkatsev** [comment]
tl;dr: you have the right idea about `selectSources`. However, I believe what you're asking for is not a `canPlayType` function but more of a `whichTechSupportsType` function which we can discuss but in another issue, please :)

`selectSources` loops through each tech and finds the top source that the earliest tech in the techOrder can play. So, if your tech order is `['hls', 'html5', 'flash']` and your sources are `[mp4, m3u8]`, notice that m3u8 is second. It will run through the sources on the `hls` tech which will choose the `m3u8`. However, this is actually slightly different than how the native source selection works (and we should have an issue about it somewhere but I cannot find it). Then, if you give videojs another source with a sourcelist of `[mp4, flv]`, it'll try to see if `hls` can play those, which it won't, then `html5`, which will play the `mp4`.

This story does get slightly better with source handlers because they bind to the existing html5 or flash techs. So, an MSE-based HLS implementation (which we're working on) will make the `html5` tech be able to say that it can play `m3u8` on MSE-supported browsers so it will choose the `html5` tech. Though, in this case, it'll end up choosing mp4 in both cases because the mp4 came before the m3u8 since `maybe` and `probably` are not taken into account.

You could force a specific tech, but again, the player level `canPlayType` should only care about whether the player can currently play a particular type or not. Source selection and tech selection are other (but related) issues.

**Comment by nickygerritsen** [comment]
OK I created #2705 for the `whichTechSupportsType` thingy.

I think then this current issue is indeed not too hard, although maybe it requires some modification of the `canHandleSource` functions to extract some function like `canHandleType`? At least both the Flash and HTML5 tech have something like that already, but they are private for now.

**Comment by gkatsev** [comment]
That's for that issue. And yeah, I don't think this one will be hard. Just surprising that it didn't exist already.

**Comment by gkatsev** [comment]
#2709 has been merged.
